narrow: Save blue box position and restore on navigation.

evykassirer opened the pull request:
This saves the blue box position as state on the location in brower history, with `history.replaceState`.

The position is restored when a narrow is activated.

Fixes #20066.


![Kapture 2023-08-08 at 18 07 22](https://github.com/zulip/zulip/assets/5634097/8846f35a-5255-4d42-a890-a179bdfcb973)

evykassirer commented:
I'm planning to wait for code review to verify this method before writing tests.

karlstolley commented:
This seems to work okay on testing. When I try racing through views really quickly, it seems to hang onto the history as expected.

timabbott commented:
I think `save_narrow_state` might better belong in the `browser_history` module; and if you do that, I expect that you won't be pressured to write tests for it. 

(I think this is probably the type of code where node tests are quite low value compared to manual testing)

evykassirer commented:
updates:

* Moved the code to `browser_history.js`
* Switched to throttling instead of debounce, as this ensures we call the function in a timely manner
* Added another call to `update_selection` which is used when manually setting the narrow position, and used `flush()` to ensure this call happens without any delay

evykassirer commented:
> I think save_narrow_state might better belong in the browser_history module; and if you do that, I expect that you won't be pressured to write tests for it.

turns out `browser_history` also wants code coverage -- should I exclude this file or would you recommend something else?

timabbott commented:
Yeah that sounds reasonable; we've already achieved the goal of testing the existing logic, and this new logic is probably not going to get a ton out of node tests.

evykassirer commented:
Okay that creates an import cycle (which I now recall dealing with earlier) -- so I think I'll just put it in its own file?

evykassirer commented:
@timabbott okay got it passing CI! back to you :)

timabbott commented on web/src/narrow.js:
```diff
@@ -792,6 +793,8 @@ export function update_selection(opts) {
     }
     unread_ops.process_visible();
+    narrow_history.save_narrow_state();
+    narrow_history.save_narrow_state.flush();
```
This kinda feels like an abstraction violation; can this be just `narrow_history.save_and_flush_narrow_state` or something like that, and the `.flush` internals of the debouncing not be exposed here?

evykassirer replied:
Yeah great point! I just created a `save_narrow_state_and_flush` function.

timabbott commented:
Posted one comment, otherwise lgtm. Thanks for all the work on this!

evykassirer commented:
Pushed with that change!

timabbott commented:
Merged, thanks @evykassirer!
